contracts/interchain-token-service/src/contract/execute/mod.rs

```diff
@@ -408,6 +424,52 @@ fn update_chain(storage: &mut dyn Storage, config: msg::ChainConfig) -> Result<(
     Ok(())
 }
 
+pub fn register_p2p_token_instance(
+    deps: DepsMut,
+    token_id: TokenId,
+    instance_chain: ChainNameRaw,
+    origin_chain: ChainNameRaw,
+    decimals: u8,
+    supply: TokenSupply,
+) -> Result<Response, Error> {
+    ensure_chain_is_registered(deps.storage, instance_chain.clone())?;
+    ensure_chain_is_registered(deps.storage, origin_chain.clone())?;
+
+    if state::may_load_token_instance(deps.storage, instance_chain.clone(), token_id)
+        .change_context(Error::State)?
+        .is_some()
+    {
+        return Err(report!(Error::TokenAlreadyDeployed {
+            token_id,
+            chain: instance_chain
+        }));
+    }
+
+    state::save_token_instance(
+        deps.storage,
+        instance_chain.clone(),
+        token_id,
+        &TokenInstance { supply, decimals },
+    )
+    .change_context(Error::State)?;
+
+    match state::may_load_token_config(deps.storage, &token_id).change_context(Error::State)? {
+        Some(token_config) => {
```

Comment:
```suggestion
        Some(TokenConfig { origin_chain: stored_origin_chain }) => {
```
could also destructure